File: libtmux/src/main/java/io/github/libtmux/query/Fields.java
package io.github.libtmux.query;

import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;
import java.util.regex.Pattern;

/**
 * Typed field handles.
 *
 * <p>A caller supplies each accessor, so no backend may trust the name without matching this exact
 * handle against a model it owns.
 *
 * <p>Each kind exposes only the operators its type supports, so {@code Pane_.index().startsWith(..)}
 * is a compile error rather than a runtime class cast. This is the half of the metamodel that has to
 * exist by hand or by generator; the generator only writes the accessors, never the semantics.
 */
public final class Fields {

    private Fields() {}

    public static <T> TextField<T> text(String id, Function<T, String> accessor) {
        return new TextField<>(new FieldRef<>(id, FieldKind.TEXT, accessor));
    }

    public static <T> NumberField<T> number(String id, Function<T, Integer> accessor) {
        return new NumberField<>(new FieldRef<>(id, FieldKind.NUMBER, accessor));
    }

    public static <T> FlagField<T> flag(String id, Function<T, Boolean> accessor) {
        return new FlagField<>(new FieldRef<>(id, FieldKind.FLAG, accessor));
    }

    public static <T, R> ToManyRef<T, R> toMany(String id, Function<T, List<R>> navigate) {
        return new ToManyRef<>(id, navigate);
    }

    public static <T, R> ToOneRef<T, R> toOne(String id, Function<T, Optional<R>> navigate) {
        return new ToOneRef<>(id, navigate);
    }

    /** String-valued field. */
    public record TextField<T>(FieldRef<T, String> ref) {

        public FilterExpr<T> is(String value) {
            return new FilterExpr.Compare<>(ref, Operator.EQUALS, value);
        }

        public FilterExpr<T> isNot(String value) {
            return new FilterExpr.Compare<>(ref, Operator.NOT_EQUALS, value);
        }

        public FilterExpr<T> contains(String value) {
            return new FilterExpr.Compare<>(ref, Operator.CONTAINS, value);
        }

        public FilterExpr<T> startsWith(String value) {
            return new FilterExpr.Compare<>(ref, Operator.STARTS_WITH, value);
        }

        public FilterExpr<T> endsWith(String value) {
            return new FilterExpr.Compare<>(ref, Operator.ENDS_WITH, value);
        }

        public FilterExpr<T> matches(Pattern value) {
            return new FilterExpr.Compare<>(ref, Operator.MATCHES, value);
        }

        public FilterExpr<T> in(Collection<String> values) {
            return new FilterExpr.Compare<>(ref, Operator.IN, List.copyOf(values));
        }
    }

    /** Integer-valued field. No text operators reach it. */
    public record NumberField<T>(FieldRef<T, Integer> ref) {

        public FilterExpr<T> is(int value) {
            return new FilterExpr.Compare<>(ref, Operator.EQUALS, value);
        }

        public FilterExpr<T> isNot(int value) {
            return new FilterExpr.Compare<>(ref, Operator.NOT_EQUALS, value);
        }

        public FilterExpr<T> lessThan(int value) {
            return new FilterExpr.Compare<>(ref, Operator.LESS_THAN, value);
        }

        public FilterExpr<T> atMost(int value) {
            return new FilterExpr.Compare<>(ref, Operator.AT_MOST, value);
        }

        public FilterExpr<T> greaterThan(int value) {
            return new FilterExpr.Compare<>(ref, Operator.GREATER_THAN, value);
        }

        public FilterExpr<T> atLeast(int value) {
            return new FilterExpr.Compare<>(ref, Operator.AT_LEAST, value);
        }
    }

    /** Boolean-valued field. */
    public record FlagField<T>(FieldRef<T, Boolean> ref) {

        public FilterExpr<T> is(boolean value) {
            return new FilterExpr.Compare<>(ref, Operator.EQUALS, value);
        }

        public FilterExpr<T> isNot(boolean value) {
            return new FilterExpr.Compare<>(ref, Operator.NOT_EQUALS, value);
        }

        public FilterExpr<T> isTrue() {
            return is(true);
        }

        public FilterExpr<T> isFalse() {
            return is(false);
        }
    }

    /** To-many relation. Quantifiers are the only way in, so an unquantified relation cannot compile. */
    public static final class ToManyRef<T, R> {

        private final String id;
        private final Function<T, List<R>> navigate;

        private ToManyRef(String id, Function<T, List<R>> navigate) {
            this.id = FieldRef.requireId(id);
            this.navigate = Objects.requireNonNull(navigate, "navigate");
        }

        public String id() {
            return id;
        }

        public Function<T, List<R>> navigate() {
            return navigate;
        }

        public FilterExpr.ToMany<T, R> any(FilterExpr<R> predicate) {
            return new FilterExpr.ToMany<>(this, FilterExpr.Quantifier.ANY, predicate);
        }

        public FilterExpr.ToMany<T, R> all(FilterExpr<R> predicate) {
            return new FilterExpr.ToMany<>(this, FilterExpr.Quantifier.ALL, predicate);
        }

        public FilterExpr.ToMany<T, R> none(FilterExpr<R> predicate) {
            return new FilterExpr.ToMany<>(this, FilterExpr.Quantifier.NONE, predicate);
        }
    }

    /** To-one relation. */
    public static final class ToOneRef<T, R> {

        private final String id;
        private final Function<T, Optional<R>> navigate;

        private ToOneRef(String id, Function<T, Optional<R>> navigate) {
            this.id = FieldRef.requireId(id);
            this.navigate = Objects.requireNonNull(navigate, "navigate");
        }

        public String id() {
            return id;
        }

        public Function<T, Optional<R>> navigate() {
            return navigate;
        }

        public FilterExpr.ToOne<T, R> is(FilterExpr<R> predicate) {
            return new FilterExpr.ToOne<>(this, predicate);
        }
    }
}

File: libtmux-kotlin/src/main/kotlin/io/github/libtmux/kotlin/query/Fields.kt
package io.github.libtmux.kotlin.query

import io.github.libtmux.query.FilterExpr
import io.github.libtmux.query.Fields
import java.util.regex.Pattern

/**
 * Kotlin-idiomatic wrappers over [Fields]' typed field handles, each holding the Java handle it
 * wraps. A value class, so this costs nothing beyond the Java object it already holds — the same
 * choice [io.github.libtmux.kotlin.Server]'s live handles reject, made here on purpose: a field
 * handle carries exactly one property, which is what makes the boxing tradeoff free.
 */
@JvmInline
public value class TextField<T : Any> internal constructor(private val java: Fields.TextField<T>) {
    public infix fun eq(value: String): FilterExpr<T> = java.`is`(value)
    public infix fun ne(value: String): FilterExpr<T> = java.isNot(value)
    public infix fun startsWith(value: String): FilterExpr<T> = java.startsWith(value)
    public infix fun endsWith(value: String): FilterExpr<T> = java.endsWith(value)
    public infix fun contains(value: String): FilterExpr<T> = java.contains(value)
    public infix fun matches(pattern: Pattern): FilterExpr<T> = java.matches(pattern)
    public infix fun oneOf(values: Collection<String>): FilterExpr<T> = java.`in`(values)
}

@JvmInline
public value class NumberField<T : Any> internal constructor(private val java: Fields.NumberField<T>) {
    public infix fun eq(value: Int): FilterExpr<T> = java.`is`(value)
    public infix fun ne(value: Int): FilterExpr<T> = java.isNot(value)
    public infix fun lessThan(value: Int): FilterExpr<T> = java.lessThan(value)
    public infix fun atMost(value: Int): FilterExpr<T> = java.atMost(value)
    public infix fun greaterThan(value: Int): FilterExpr<T> = java.greaterThan(value)
    public infix fun atLeast(value: Int): FilterExpr<T> = java.atLeast(value)
}

@JvmInline
public value class FlagField<T : Any> internal constructor(private val java: Fields.FlagField<T>) {
    public fun isTrue(): FilterExpr<T> = java.isTrue()
    public fun isFalse(): FilterExpr<T> = java.isFalse()
}

@JvmInline
public value class ToManyField<T : Any, R : Any> internal constructor(private val java: Fields.ToManyRef<T, R>) {
    public infix fun any(predicate: FilterExpr<R>): FilterExpr<T> = java.any(predicate)
    public infix fun all(predicate: FilterExpr<R>): FilterExpr<T> = java.all(predicate)
    public infix fun none(predicate: FilterExpr<R>): FilterExpr<T> = java.none(predicate)
}

@JvmInline
public value class ToOneField<T : Any, R : Any> internal constructor(private val java: Fields.ToOneRef<T, R>) {
    public infix fun matching(predicate: FilterExpr<R>): FilterExpr<T> = java.`is`(predicate)
}
